notifications/hub.go

```diff
@@ -0,0 +1,61 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2016 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package notifications
+
+import (
+	"sync"
+)
+
+// A Hub allows subscribers to receive notifications.
+type Hub struct {
+	sync.Mutex
+	subscribers Subscribers
+}
+
+// Subscribe registers a subscriber to receive notifications.
+func (h *Hub) Subscribe(s *Subscriber) {
+	h.Lock()
+	defer h.Unlock()
+
+	h.subscribers = append(h.subscribers, s)
+}
+
+// Unsubscribe unregisters a subscriber from notifications.
+func (h *Hub) Unsubscribe(s *Subscriber) {
+	h.Lock()
+	defer h.Unlock()
+
+	for i, sub := range h.subscribers {
+		if s == sub {
+			h.subscribers = append(h.subscribers[:i], h.subscribers[i+1:]...)
+			break
```

Comment:
Not sure how relevant this is in practise, this break will mean that if there are multiple identical subscribers only the first one gets removed. Is that what we expect from the API?

I guess it comes down to what the semantic should be if someone runs `s := NewSubscriber(); Hub.Subscribe(s); Hub.Subscribe(s);` be. I.e. do we allow the same subscriber to subscribe twice? I will continue reading the rest of the pull-request, maybe my question answers itself :)